### PR TITLE
Add support for UCRT MinGW environment

### LINKED ISSUE #3054: Windows GCC 13.2 compilation failed
### Prerequisite Checklist

- [X] I searched for [existing issues](https://github.com/search?q=repo%3ASFML%2FSFML&type=issues) to prevent duplicates
- [X] I searched for [existing discussions on the forum](https://www.google.com/search?q=site%3Ahttps%3A%2F%2Fen.sfml-dev.org) to prevent duplicates
- [X] I am here to report an issue and not to just ask a question or look for help (use the [forum](https://en.sfml-dev.org/forums/index.php#c3) or [Discord](https://discord.gg/nr4X7Fh) instead)

### Describe your issue here

Compilation failed during linking

### Your Environment

- OS / distro / window manager: Windows 64bits
- SFML version: 2.6.1
- Compiler / toolchain: gcc 13.2 seh posix urct
- Special compiler / CMake flags: Release

### Steps to reproduce

```ps
$repoUrl = "https://github.com/SFML/SFML"
$gitTag = "2.6.1"


# Clone the repository at the specified tag
git clone -b $gitTag $repoUrl

# Change the working directory to external_deps/SFML
cd SFML
md build
cd build
$SFML_build = $PWD

# Execute the cmake command
cmake -G "MinGW Makefiles" -DBUILD_SHARED_LIBS=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX="$installPath" ..

# Build the project
cmake --build . --target all

# Install the project
cmake --install .
```

### Expected behavior

Compilation success

### Actual behavior

[ 15%] Built target sfml-system
[ 17%] Built target sfml-main
[ 42%] Built target sfml-window
[ 53%] Built target sfml-network
[ 54%] Linking CXX shared library ..\..\..\lib\sfml-graphics-2.dll
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: .../external_deps/SFML/extlibs/libs-mingw/x64/libfreetype.a(sfnt.c.obj):sfnt.c:(.text+0x5614): undefined reference to `_setjmp'
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: .../external_deps/SFML/extlibs/libs-mingw/x64/libfreetype.a(smooth.c.obj):smooth.c:(.text+0x77a): undefined reference to `_setjmp'
collect2.exe: error: ld returned 1 exit status
mingw32-make[2]: *** [src\SFML\Graphics\CMakeFiles\sfml-graphics.dir\build.make:602: lib/sfml-graphics-2.dll] Error 1
mingw32-make[1]: *** [CMakeFiles\Makefile2:307: src/SFML/Graphics/CMakeFiles/sfml-graphics.dir/all] Error 2
mingw32-make: *** [Makefile:155: all] Error 2

### PR DESCRIPTION
**Pull request by vittorioromeo**
Fixes #3054, #2289, and more older reports on the forums. No plans to support UCRT32 or SFML 2.x. 

FLAC version is 1.3.4.
Freetype version is 2.5.5.

This is intended as a stopgap measure until dependencies are built from source, I do not intend this change to be a permanent solution.

### CODE REVIEW

**Comment by eXpl0it3r** [comment]
> No plans to support UCRT32 or SFML 2.x.

By that you mean, you're not planning on doing it?

Because we certainly will be also supporting UCRT32 and likely backport it to SFML 2

**Comment by vittorioromeo** [comment]
> By that you mean, you're not planning on doing it?

Not in this PR, at the very least. I can consider creating a separate PR to backport UCRT64 support to 2.x, though.

Generally not interested in supporting 32-bit platforms, but could change my mind if there's high demand.

**Comment by vittorioromeo** [comment]
SFMLBuildMaster: Build this please.

File: cmake/Config.cmake [hunk 1]
```diff
@@ -87,11 +87,21 @@ if(MSVC)
     endif()
 elseif(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
     set(SFML_COMPILER_CLANG 1)
+
+    execute_process(COMMAND "${CMAKE_CXX_COMPILER}" "-v" ERROR_VARIABLE CLANG_COMPILER_VERSION OUTPUT_VARIABLE CLANG_COMPILER_VERSION)
```
Comment (eXpl0it3r) [inline cmake/Config.cmake:91]: Nitpick: I'd order it the other way around, first `OUTPUT_VARIABLE` then `ERROR_VARIBLE`

File: cmake/Config.cmake [hunk 2]
```diff
@@ -87,11 +87,21 @@ if(MSVC)
...
+
+    if("${CLANG_COMPILER_VERSION}" MATCHES "ucrt")
+        set(SFML_UCRT 1)
```
Comment (eXpl0it3r) [inline cmake/Config.cmake:94]: I'd use something less "generic". Like `SFML_RUNTIME_UCRT`

File: cmake/Config.cmake [hunk 3]
```diff
@@ -87,11 +87,21 @@ if(MSVC)
...
+    endif()
 elseif(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
     set(SFML_COMPILER_GCC 1)
 
-    execute_process(COMMAND "${CMAKE_CXX_COMPILER}" "--version" OUTPUT_VARIABLE GCC_COMPILER_VERSION)
+    execute_process(COMMAND "${CMAKE_CXX_COMPILER}" "-v" ERROR_VARIABLE GCC_COMPILER_VERSION OUTPUT_VARIABLE GCC_COMPILER_VERSION)
     string(REGEX MATCHALL ".*(tdm[64]*-[1-9]).*" SFML_COMPILER_GCC_TDM "${GCC_COMPILER_VERSION}")
```
Comment (eXpl0it3r) [inline cmake/Config.cmake:100]: I double checked this, this should still work ✅

File: src/SFML/CMakeLists.txt [hunk 4]
```diff
@@ -6,8 +6,14 @@ if(SFML_OS_WINDOWS)
     list(APPEND CMAKE_INCLUDE_PATH "${PROJECT_SOURCE_DIR}/extlibs/headers")
     if(SFML_COMPILER_GCC OR (SFML_COMPILER_CLANG AND MINGW))
         if(ARCH_32BITS)
+            if(SFML_UCRT)
+                message(FATAL_ERROR "UCRT32 environment is not supported")
+            endif()
```
Comment (eXpl0it3r) [inline src/SFML/CMakeLists.txt:11]: I'm of the opinion that we should just do it. If you could build the other libraries, then this is done in a breeze by installing a 32-bits compiler.

**Review by eXpl0it3r — CHANGES_REQUESTED** [review]
Wanted to updated the PR directly, but was held up by other things

File: cmake/Config.cmake [hunk 5]
```diff
@@ -110,13 +110,25 @@ elseif(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
 
     execute_process(COMMAND "${CMAKE_CXX_COMPILER}" "--version" OUTPUT_VARIABLE CLANG_VERSION_OUTPUT)
     string(REGEX REPLACE ".*clang version ([0-9]+\\.[0-9]+).*" "\\1" SFML_CLANG_VERSION "${CLANG_VERSION_OUTPUT}")
+
+    execute_process(COMMAND "${CMAKE_CXX_COMPILER}" "-v" OUTPUT_VARIABLE CLANG_COMPILER_VERSION ERROR_VARIABLE CLANG_COMPILER_VERSION)
+
+    if("${CLANG_COMPILER_VERSION}" MATCHES "ucrt")
+        set(SFML_RUNTIME_UCRT 1)
+    endif()
```
Comment (ChrisThrasher) [inline cmake/Config.cmake:118]: I've only heard of UCRT MinGW. Is there also a URCT version of Clang someone may use?

**Review by ChrisThrasher — APPROVED** [review]
I have no means to test this so if you think it works then it LGTM.

File: cmake/Config.cmake (hunk 5 above)
Comment (eXpl0it3r) [reply cmake/Config.cmake:118]: MinGW is more like the toolchain, within which you can have GCC or Clang as compiler and they can link the universal C runtime (UCRT) or Microsoft Visual Studio C runtime (MSVCRT)

**Comment by eXpl0it3r** [comment]
I've built SFML's example with the specific compilers on Windows and it picks the libs up correctly and the example run fine.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Review by eXpl0it3r — APPROVED** [review]

**Comment by eXpl0it3r** [comment]
I believe we missed to update the install rules:

```cmake
    if(NOT SFML_USE_SYSTEM_DEPS)
        # install the binaries of SFML dependencies
        if(ARCH_ARM64)
            install(DIRECTORY extlibs/libs-msvc-universal/ARM64/ DESTINATION ${CMAKE_INSTALL_LIBDIR})
        elseif(ARCH_X86)
            if(SFML_COMPILER_MSVC OR (SFML_COMPILER_CLANG AND NOT MINGW))
                install(DIRECTORY extlibs/libs-msvc-universal/x86/ DESTINATION ${CMAKE_INSTALL_LIBDIR})
            else()
                install(DIRECTORY extlibs/libs-mingw/x86/ DESTINATION ${CMAKE_INSTALL_LIBDIR})
            endif()
        elseif(ARCH_X64)
            if(SFML_COMPILER_MSVC OR (SFML_COMPILER_CLANG AND NOT MINGW))
                install(DIRECTORY extlibs/libs-msvc-universal/x64/ DESTINATION ${CMAKE_INSTALL_LIBDIR})
            else()
                install(DIRECTORY extlibs/libs-mingw/x64/ DESTINATION ${CMAKE_INSTALL_LIBDIR})
            endif()
        endif()
    endif()
```

**Pull request merged**